Mathlib/Topology/UniformSpace/Basic.lean

```diff
@@ -0,0 +1,1966 @@
+First note that the triangle inequality in a metric space is equivalent to
+`∀ (x y z : X) (r r' : ℝ), dist x y ≤ r → dist y z ≤ r' → dist x z ≤ r + r'`.
+Then, for any `V` and `W` with type `Set (X × X)`, the composition `V ○ W : Set (X × X)` is
+defined as `{ p : X × X | ∃ z, (p.1, z) ∈ V ∧ (z, p.2) ∈ W }`.
+In the metric space case, if `V = { p | dist p.1 p.2 ≤ r }` and `W = { p | dist p.1 p.2 ≤ r' }`
+then the triangle inequality, as reformulated above, says `V ○ W` is contained in
+`{p | dist p.1 p.2 ≤ r + r'}` which is the entourage associated to the radius `r + r'`.
+In general we have `mem_ball_comp (h : y ∈ ball x V) (h' : z ∈ ball y W) : z ∈ ball x (V ○ W)`.
+Note that this discussion does not depend on any axiom imposed on the uniformity filter,
+it is simply captured by the definition of composition.
+
+The uniform space axioms ask the filter `𝓤 X` to satisfy the following:
+* every `V ∈ 𝓤 X` contains the diagonal `idRel = { p | p.1 = p.2 }`. This abstracts the fact
+  that `dist x x ≤ r` for every non-negative radius `r` in the metric space case and also that
+  `x - x` belongs to every neighborhood of zero in the topological group case.
+* `V ∈ 𝓤 X → Prod.swap '' V ∈ 𝓤 X`. This is tightly related the fact that `dist x y = dist y x`
+  in a metric space, and to continuity of negation in the topological group case.
+* `∀ V ∈ 𝓤 X, ∃ W ∈ 𝓤 X, W ○ W ⊆ V`. In the metric space case, it corresponds
+  to cutting the radius of a ball in half and applying the triangle inequality.
+  In the topological group case, it comes from continuity of addition at `(0, 0)`.
+
+These three axioms are stated more abstractly in the definition below, in terms of
+operations on filters, without directly manipulating entourages.
+
+## Main definitions
+
+* `UniformSpace X` is a uniform space structure on a type `X`
+* `UniformContinuous f` is a predicate saying a function `f : α → β` between uniform spaces
+  is uniformly continuous : `∀ r ∈ 𝓤 β, ∀ᶠ (x : α × α) in 𝓤 α, (f x.1, f x.2) ∈ r`
+
+In this file we also define a complete lattice structure on the type `UniformSpace X`
+of uniform structures on `X`, as well as the pullback (`UniformSpace.comap`) of uniform structures
+coming from the pullback of filters.
+Like distance functions, uniform structures cannot be pushed forward in general.
+
+## Notations
+
+Localized in `uniformity`, we have the notation `𝓤 X` for the uniformity on a uniform space `X`,
+and `○` for composition of relations, seen as terms with type `Set (X × X)`.
+
+## Implementation notes
+
+There is already a theory of relations in `data/rel.lean` where the main definition is
+`def Rel (α β : Type*) := α → β → Prop`.
+The relations used in the current file involve only one type, but this is not the reason why
+we don't reuse `data/rel.lean`. We use `Set (α × α)`
+instead of `Rel α α` because we really need sets to use the filter library, and elements
+of filters on `α × α` have type `Set (α × α)`.
```

Comment:
```suggestion
we don't reuse `Data/Rel.lean`. We use `Set (α × α)`
instead of `Rel α α` because we really need sets to use the filter library, and elements
of filters on `α × α` have type `Set (α × α)`.
```